File: src/Flare.Api/Model/LogAggregateRequest.cs
using MemoryPack;

namespace Flare.Api.Model;

/// <summary>Optional secondary grouping dimension for <c>POST /api/logs/aggregate</c>.</summary>
public enum LogAggregateGroupBy
{
    None,
    Service,
    Level,

    /// <summary>
    /// One attribute key's value - <see cref="LogAggregateRequest.GroupByAttributeBag"/> +
    /// <see cref="LogAggregateRequest.GroupByAttributeKey"/> name which. Appended last so the
    /// existing members keep their MemoryPack ordinals.
    /// </summary>
    Attribute,

    /// <summary>OTel instrumentation scope name (<c>ScopeName</c> - the .NET logger category). Appended after <see cref="Attribute"/> for the same ordinal-stability reason.</summary>
    Scope,
}

/// <summary>Request body for <c>POST /api/logs/aggregate</c> - volume-over-time chart data.</summary>
[MemoryPackable]
public sealed partial record LogAggregateRequest
{
    /// <summary>See <see cref="LogSearchRequest.Filter"/>'s doc comment - the same JSON-deserialization caveat applies here.</summary>
    public LogFilter Filter { get; init; } = new();

    /// <summary>Bucket width, e.g. 60 for 1-minute buckets. Compiles to <c>toStartOfInterval(Timestamp, INTERVAL n SECOND)</c>.</summary>
    public required int BucketWidthSeconds { get; init; }

    public LogAggregateGroupBy GroupBy { get; init; } = LogAggregateGroupBy.None;

    /// <summary>
    /// Optional chain of app-side transforms applied, in list order, to each returned
    /// bucket's <see cref="LogAggregateBucket.Count"/> after the ClickHouse query runs -
    /// see <see cref="Query.LogPostProcessor"/>. Null/empty = no post-processing (default).
    /// The Logs-explorer counterpart to <c>MetricQueryRequest.PostProcessFunctions</c>
    /// (ADR-0038/0039) - appended last, same append-only field-versioning convention that
    /// pair's own remarks document. When <see cref="GroupBy"/> is set, each distinct
    /// <see cref="LogAggregateBucket.GroupKey"/>'s buckets are processed as their own
    /// independent series, not run together as one - see <see cref="Query.LogPostProcessor"/>'s
    /// remarks.
    /// </summary>
    public IReadOnlyList<LogPostProcessFunction>? PostProcessFunctions { get; init; }

    /// <summary>
    /// Which attribute bag <see cref="GroupByAttributeKey"/> is looked up in when
    /// <see cref="GroupBy"/> is <see cref="LogAggregateGroupBy.Attribute"/>; ignored otherwise.
    /// Appended after <see cref="PostProcessFunctions"/> - same append-only versioning.
    /// </summary>
    public AttributeBag GroupByAttributeBag { get; init; } = AttributeBag.Log;

    /// <summary>
    /// Attribute key to group by when <see cref="GroupBy"/> is
    /// <see cref="LogAggregateGroupBy.Attribute"/> - required (non-blank) in that case,
    /// ignored otherwise. Only the <see cref="Query.LogAggregateQueryBuilder.AttributeGroupLimit"/>
    /// most frequent values in the window get their own series; every other value's events
    /// come back under a null <see cref="LogAggregateBucket.GroupKey"/> ("other"). Events
    /// missing the key group under an empty-string key, so the stacked total still matches
    /// the ungrouped chart.
    /// </summary>
    public string? GroupByAttributeKey { get; init; }
}

/// <summary>
/// Which per-query post-processing transform a <see cref="LogPostProcessFunction"/> applies -
/// the Logs-explorer counterpart to <c>MetricPostProcessFunctionType</c> (ADR-0038/0039),
/// shipped for Logs by ADR-0041. Same member set and ordinals - see
/// <see cref="Query.LogPostProcessor"/>'s remarks for how the semantics diverge from the
/// metrics version now that there's no nullable <see cref="LogAggregateBucket.Count"/> slot
/// to carry a "gap" through. MemoryPack encodes this as its numeric ordinal, so existing
/// members must keep their ordinals if new ones are ever appended.
/// </summary>
public enum LogPostProcessFunctionType
{
    ClampMin,
    ClampMax,
    Absolute,
    Log2,
    Log10,
    CumulativeSum,
    EwmaSmoothing,
    MedianSmoothing,
}

/// <summary>
/// One step in a <see cref="LogAggregateRequest.PostProcessFunctions"/> chain, applied in
/// list order (each function's output feeds the next) by <see cref="Query.LogPostProcessor"/>.
/// Field-for-field identical shape to <c>MetricPostProcessFunction</c>.
/// </summary>
[MemoryPackable]
[GenerateTypeScript]
public sealed partial record LogPostProcessFunction
{
    public required LogPostProcessFunctionType Type { get; init; }

    /// <summary>Threshold for <see cref="LogPostProcessFunctionType.ClampMin"/>/<see cref="LogPostProcessFunctionType.ClampMax"/> - required for those two, ignored (may be null) for every other function.</summary>
    public double? Value { get; init; }

    /// <summary>
    /// Trailing lookback, in buckets, for <see cref="LogPostProcessFunctionType.EwmaSmoothing"/>
    /// (converted to a decay factor via the standard <c>alpha = 2 / (N + 1)</c> N-period
    /// formula) / <see cref="LogPostProcessFunctionType.MedianSmoothing"/> (a literal
    /// trailing window of up to N points) - required (&gt;= 1) for those two, ignored (may
    /// be null) for every other function.
    /// </summary>
    public int? WindowSize { get; init; }
}

/// <summary>
/// One bucketed value. <see cref="GroupKey"/> is null when <see cref="LogAggregateGroupBy.None"/>
/// was requested, or - for <see cref="LogAggregateGroupBy.Attribute"/> - for the rolled-up
/// "other" series (see <see cref="LogAggregateRequest.GroupByAttributeKey"/>). <see cref="Count"/> is <c>double</c> (not <c>long</c>) so the same shape
/// can carry a SQL-query-row <c>avg()</c>/<c>sum()</c> result (see
/// <c>Query.LogQl.LogQlQueryBuilder</c>) as well as this endpoint's own always-integral
/// <c>count()</c> - a whole-number value still round-trips through JSON exactly (e.g. `25`,
/// not `25.0`), so this is a no-op for every existing <c>/api/logs/aggregate</c> caller.
/// </summary>
[MemoryPackable]
public sealed partial record LogAggregateBucket
{
    public required DateTimeOffset BucketStart { get; init; }

    public string? GroupKey { get; init; }

    public required double Count { get; init; }
}

/// <summary>Response body for <c>POST /api/logs/aggregate</c>.</summary>
[MemoryPackable]
public sealed partial record LogAggregateResponse
{
    public required IReadOnlyList<LogAggregateBucket> Buckets { get; init; }
}

File: src/dashboard/src/lib/memorypack/LogAggregateRequest.ts
// Hand-written companion to MemoryPack's generated TypeScript classes - NOT itself
// generated. Mirrors `src/Flare.Api/Model/LogAggregateRequest.cs`'s `LogAggregateRequest`
// field-for-field, in declared order. Can't carry `[GenerateTypeScript]` itself because it
// nests `LogFilter` (blocked - see `$lib/memorypack/LogFilter.ts`'s header comment).
// `groupBy` is a raw MemoryPack numeric ordinal (converted to string at `api.ts`'s module
// boundary via `$lib/memorypack/enums.ts`'s `logAggregateGroupByFromString`).
// `postProcessFunctions` was appended after every pre-existing field for ADR-0041's
// post-processing chain, same versioning convention `MetricQueryRequest.ts`'s own
// `postProcessFunctions` append documents - `LogPostProcessFunction` itself has no
// DateTimeOffset/nesting problem, so it's a real generated class
// (`$lib/generated/memorypack/LogPostProcessFunction.js`), reused here directly.
// `groupByAttributeBag`/`groupByAttributeKey` were appended after that, same convention -
// the bag is a raw `AttributeBag` ordinal (`$lib/memorypack/enums.ts`'s
// `attributeBagFromString`).

import { MemoryPackWriter } from '$lib/generated/memorypack/MemoryPackWriter.js';
import { MemoryPackReader } from '$lib/generated/memorypack/MemoryPackReader.js';
import { LogFilter } from '$lib/memorypack/LogFilter';
import { LogPostProcessFunction } from '$lib/generated/memorypack/LogPostProcessFunction.js';

export class LogAggregateRequest {
	filter: LogFilter | null;
	bucketWidthSeconds: number;
	groupBy: number;
	postProcessFunctions: (LogPostProcessFunction | null)[] | null;
	groupByAttributeBag: number;
	groupByAttributeKey: string | null;

	constructor() {
		this.filter = null;
		this.bucketWidthSeconds = 0;
		this.groupBy = 0;
		this.postProcessFunctions = null;
		this.groupByAttributeBag = 0;
		this.groupByAttributeKey = null;
	}

	static serialize(value: LogAggregateRequest | null): Uint8Array {
		const writer = MemoryPackWriter.getSharedInstance();
		this.serializeCore(writer, value);
		return writer.toArray();
	}

	static serializeCore(writer: MemoryPackWriter, value: LogAggregateRequest | null): void {
		if (value == null) {
			writer.writeNullObjectHeader();
			return;
		}

		writer.writeObjectHeader(6);
		LogFilter.serializeCore(writer, value.filter);
		writer.writeInt32(value.bucketWidthSeconds);
		writer.writeInt32(value.groupBy);
		writer.writeArray(value.postProcessFunctions, (writer, x) => LogPostProcessFunction.serializeCore(writer, x));
		writer.writeInt32(value.groupByAttributeBag);
		writer.writeString(value.groupByAttributeKey);
	}

	static deserialize(buffer: ArrayBuffer): LogAggregateRequest | null {
		return this.deserializeCore(new MemoryPackReader(buffer));
	}

	static deserializeCore(reader: MemoryPackReader): LogAggregateRequest | null {
		const [ok, count] = reader.tryReadObjectHeader();
		if (!ok) {
			return null;
		}

		const value = new LogAggregateRequest();
		if (count == 6) {
			value.filter = LogFilter.deserializeCore(reader);
			value.bucketWidthSeconds = reader.readInt32();
			value.groupBy = reader.readInt32();
			value.postProcessFunctions = reader.readArray((reader) => LogPostProcessFunction.deserializeCore(reader));
			value.groupByAttributeBag = reader.readInt32();
			value.groupByAttributeKey = reader.readString();
		} else if (count > 6) {
			throw new Error("Current object's property count is larger than type schema, can't deserialize about versioning.");
		} else {
			if (count == 0) return value;
			value.filter = LogFilter.deserializeCore(reader);
			if (count == 1) return value;
			value.bucketWidthSeconds = reader.readInt32();
			if (count == 2) return value;
			value.groupBy = reader.readInt32();
			if (count == 3) return value;
			value.postProcessFunctions = reader.readArray((reader) => LogPostProcessFunction.deserializeCore(reader));
			if (count == 4) return value;
			value.groupByAttributeBag = reader.readInt32();
			if (count == 5) return value;
			value.groupByAttributeKey = reader.readString();
			if (count == 6) return value;
		}
		return value;
	}
}
